### PR TITLE
feature: setup manuallb with GCE VMs for ABM

### LINKED ISSUE #226: Improve ABM on GCE Terraform guide to enable manual-lb deployment
### Description
- The current [anthos-bm-gcp-terraform](https://github.com/GoogleCloudPlatform/anthos-samples/tree/main/anthos-bm-gcp-terraform) contains two installation modes
- One of them is a two step process where the user runs the terraform script to create the GCE environment and a few additional steps to SSH into the created GCE to install ABM
- The other is an all-in-one install where running the terraform script not only creates the GCE environment but also triggers the ABM installation
- Both of these modes lead up to ABM being installed using the `bundled loadbalancer` mode. 
- What this means is that whilst it emulates a truly baremetal environment, we cannot use this setup for demo purposes because we cannot reach the workloads running inside the cluster
- Thus, we need to add support for being able to hit the workloads inside the ABM cluster running inside the GCE VMs

### What to be done?
- We update this sample to include an additional installation mode which enables setting up ABM using ManualLB mode
- We set a GCP loadbalancer as the manual LB used by ABM
- What this means is that once the installation is complete the control plane of the ABM cluster can be reached via a public IP
- In addition the workloads running inside the ABM cluster can also be reached via a public IP that routes traffic to the ingress controller setup in ABM by default

### PR DESCRIPTION
**Pull request by Shabirmean**
#### Description
- A detailed description can be found in issue #226 
- We add an additional installation mode that installs ABM in the `manual loadbalancer` mode
- We use a GCP load balancer for this
- We also ensure the ingress controller inside the ABM cluster can be reached via a public load balancer as well 

#### Change summary
- Add a new terraform module named `loadbalancer`
- The module supports the creation of HTTP and TCP loadbalancer
- Update the `external-ip` module to support creation of a gloabl ip address
- Update the main script to support an additional installation mode called `manuallb`
- Update the main script to create 2 loadbalancers using the `loadbalancer` module when the installation mode is `manuallb`
- We create 2 loadbalancers because one is for the controlplane and the other for the ingress of the ABM cluster
- Update the outputs of the main scripts to show the external IP addresses of both loadbalancer
- Update the READMEs of all the modules and main script to include the new changes
- Update the unit tests for the main script to match the new changes
- Add additional unit tests for the newly created loadbalancer module
- Add `terraform-docs` github actions to the CI files to ensure that changes to variables and module definitions are captured on new PRs.

#### Key notes for reviewers:
- The `CLA Check` is failing ❌ because whilst setting up the [terraform-docs](https://github.com/terraform-docs/gh-actions) Github action a few commits were added by that action-bot. The existence of those commits is what's causing it.

#### Related PRs/Issues
- N/A

### CODE REVIEW

File: .github/terraform-docs/module.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Copyright 2022 Google LLC
```
Comment (ckim328) [inline .github/terraform-docs/module.yaml:1]: thanks for getting this ready @Shabirmean! 
Quick question : for the terraform-docs, i see that there is both a `module.yaml` and `main.yaml` in the .github folder. Is it possible to combine them, or do they need to be separated?

Comment (ckim328) [reply .github/terraform-docs/module.yaml:1]: oh nevermind, i see the flow!

File: .github/gh_runner_dependencies.sh [hunk 2]
```diff
@@ -25,12 +35,15 @@ echo "export GOPATH=$HOME/go" >> "$HOME"/.profile
...
+# install the addlicense tool used to check files for license headers
 go get -u github.com/google/addlicense
 sudo ln -s "$HOME"/go/bin/addlicense /bin
 
+# install the tfenv tool to manage terraform versions
 git clone https://github.com/tfutils/tfenv.git ~/.tfenv
 sudo ln -s "$HOME"/.tfenv/bin/* /bin
 
+# install the terraform versions and configure it to the latest
 tfenv install 0.14.9
```
Comment (ckim328) [inline .github/gh_runner_dependencies.sh:47]: nit: do we need all the `tfenv` installs? especially for the ones `>v.1.0.0`

Comment (Shabirmean) [reply .github/gh_runner_dependencies.sh]: Fixed in https://github.com/GoogleCloudPlatform/anthos-samples/pull/218/commits/a0a7734212520f846a17c69dd54bb4beb37c673b

File: anthos-bm-gcp-terraform/docs/manuallb_install.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+This guide is an extension to the [all-in-one install guide](one_click_install.md).
+The `all-in-one install` is a single run installation which automatically
+triggers the complete **Anthos on bare metal** installation when the Terraform
+script is run. The default installation sets up Anthos on bare metal to run
+using a [bundled loadbalancer](https://cloud.google.com/anthos/clusters/docs/bare-metal/latest/installing/bundled-lb).
+This means that the loadbalancer and the cluster services are reachable only
+from inside one of the GCE VMs _(i.e. Admin workstation, Controlplane nodes, Worker nodes)_
+created during the installation process.
+
+While, this setup closely represents how **Anthos on bare metal (ABM)** will be
```
Comment (ckim328) [inline anthos-bm-gcp-terraform/docs/manuallb_install.md:12]: nit: ' While this setup'

Comment (Shabirmean) [reply anthos-bm-gcp-terraform/docs/manuallb_install.md]: Fixed in: 81e810a86a4cc635b10cc2de13a410a55ef56d21

File: anthos-bm-gcp-terraform/docs/manuallb_install.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+You can find your cluster's `kubeconfig` file on the admin machine in the
+`bmctl-workspace` directory _(see logs of the installation for the exact path)_.
+Since the `API Server` of the cluster is exposed via the GCP Loadbalancer, the
+`kubeconfig` file is configured to use the **Public IP Address** of the
+loadbalancer. Thus, you can copy the `kubeconfig` file to your local workstation
+and interact with the cluster using `kubectl`.
+
+To verify your deployment, complete the following steps:
+
+1. Copy the contents of `kubeconfig` file to your workstation.
```
Comment (ckim328) [inline anthos-bm-gcp-terraform/docs/manuallb_install.md:228]: i had to set my project id in my workstation, could be worth noting here?

Comment (Shabirmean) [reply anthos-bm-gcp-terraform/docs/manuallb_install.md:228]: [These lines](https://github.com/GoogleCloudPlatform/anthos-samples/pull/218/files#diff-dbd4f6dc96e4f99879f1ec3afb8df3d8024fc08a37098ebb1dae62ec0e5816c3R234-R238) have an instruction for it. Did you still face an issues?

File: anthos-bm-gcp-terraform/docs/manuallb_install.md [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+resources required for the Loadbalancers and finally triggers the installation
+of Anthos on bare metal on the provisioned VMs. The installation is triggered
+from inside the _admin workstation_ VM.
+
+Upon completion the Terraform script will print the following output to the console.
+
+> **Note:** You will notice that the output of the terraform script also has the
+> **Public IP addresses** of the `Controlplane` and `Ingress` loadbalancers. You
+> must wait until the Anthos on bare cluster installation is complete before
+> being able to use them.
```
[the remark this answers is not in the document]
Comment (Shabirmean) [reply anthos-bm-gcp-terraform/docs/manuallb_install.md]: Updated in 999c147fc7afe4189a0dc591de8bacf1fcd2cd2d

File: anthos-bm-gcp-terraform/docs/manuallb_install.md (hunk 4 above)
Comment (Shabirmean) [reply anthos-bm-gcp-terraform/docs/manuallb_install.md:229]: I had linked to the lines `234-238` on the manuallb_install.md file. If you click the link and leave it a **few seconds** then it should scroll down to the exact highlighted spot. It ain't working?

File: anthos-bm-gcp-terraform/docs/images/manuallb_architecture.svg [hunk 6]
```diff
@@ -0,0 +1 @@
...-0.296875 -0.359375 -0.828125 -0.359375zm7.641815 3.171875l1.890625 -5.46875l0.953125 0l-2.453125 6.625l-0.78125 0l-2.453125 -6.625l0.953125 0l1.890625 5.46875zm4.872345 -5.46875l2.15625 5.40625l2.171875 -5.40625l1.140625 0l0 6.625l-0.875 0l0 -2.578125l0.078125 -2.796875l-2.171875 5.375l-0.671875 0l-2.171875 -5.359375l0.078125 2.78125l0 2.578125l-0.875 0l0 -6.625l1.140625 0z" fill-rule="nonzero"/></g><path fill="#000000" fill-opacity="0.0" d="m250.4892 743.60394l22.729645 0l0 22.729675l-22.729645 0z" fill-rule="evenodd"/><g transform="matrix(0.32013595800524935 0.0 0.0 0.32013595800524935 250.4891939632546 743.6039627296587)"><clipPath id="p.40"><path d="m0 0l71.0 0l0 71.0l-71.0 0z" clip-rule="evenodd"/></clipPath><image clip-path="url(#p.40)" fill="#000" width="71.0" height="71.0" x="0.0" y="0.0" preserveAspectRatio="none" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEcAAABHCAYAAABVsFofAAADWklEQVR4Xu2bzWoTURTH+wIK6gv4Di7dJYJKwWW3gkhCXRVcdGGXNqkWCknBxtCmG3eVtEEXiU0LiviVuLAUfJEk8xUY50xyp/HMzJk7yWSaZs4ffrszc+78KJ2TmTsLCxEku99bzez1TC+yFXUT14tk9tQVXO9Q6RVwvUimoi676p3jlBKuv9SwHCIshwjLIcJyiLAcIiyHCMshwnKIJF5OrWXcrbaMF168aWr1jQ+q6cX2sf4F1wtKTeMjrneOa2jfcL1g50Sr4XpBsaH/xPWCWr
```
Comment (ckim328) [inline anthos-bm-gcp-terraform/docs/images/manuallb_architecture.svg:1]: Amazing job this diagram @Shabirmean 👏 👏 👏 👏 👏 👏 👏 👏 👏 👏

**Pull request merged**